aioftp/client.py

```diff
@@ -547,13 +697,38 @@ def list(self, path="", *, recursive=False):
 
             >>> stats = await client.list()
         """
-        class AsyncClientLister(AsyncListerMixin):
+        class AsyncListerClient(AsyncListerMixin):
 
             async def _new_stream(cls, local_path):
 
                 cls.path = local_path
-                command = str.strip("MLSD " + str(cls.path))
-                return await self.get_stream(command, "1xx")
+                str_path = " " + str(cls.path)
+                if self.list_compatibility_level > 0:
+
+                    cls._anext = cls._anext_MSLD_LIST
+
+                    if self.list_compatibility_level == 2:
+
+                        cls.process_line = self.parse_mlsx_line
+
+                        try:
+                            command = str.strip("MLSD" + str_path)
+                            return await self.get_stream(command, "1xx")
+
+                        except errors.StatusCodeError as e:
+
+                            if e.received_codes[-1] == "500":
+
+                                self.list_compatibility_level = 1
+
+                    if self.list_compatibility_level == 1:
+
+                        cls.process_line = self.parse_list_line
+                        command = str.strip("LIST" + str_path)
+                        return await self.get_stream(command, "1xx")
+                else:
+                    # TODO: implement NLST/CWD fallback
+                    pass
```

Comment:
This is not the place, where we should count performance, I think. Less state is always better. Nothing wrong to call `MLSD` and get `502` here.
> Do you mean to not implement NLST/CWD?
Nope, you can implement it. But will it be usable? Is there ftp server, which have no `MLSD` and no `LIST` support?